src/aws_iot_mqtt_client.c

```diff
@@ -212,10 +240,13 @@ IoT_Error_t aws_iot_mqtt_init(AWS_IoT_Client *pClient, IoT_Client_Init_Params *p
 	}
 	rc = aws_iot_thread_mutex_init(&(pClient->clientData.tls_read_mutex));
 	if(SUCCESS != rc) {
+		(void)aws_iot_thread_mutex_destroy(&(pClient->clientData.state_change_mutex));
 		FUNC_EXIT_RC(rc);
 	}
 	rc = aws_iot_thread_mutex_init(&(pClient->clientData.tls_write_mutex));
 	if(SUCCESS != rc) {
+		(void)aws_iot_thread_mutex_destroy(&(pClient->clientData.tls_read_mutex));
+		(void)aws_iot_thread_mutex_destroy(&(pClient->clientData.state_change_mutex));
 		FUNC_EXIT_RC(rc);
 	}
 #endif
```

Comment:
Line 263 also contains an exit on failure, should also destroy mutexes there.

Reply:
I am not sure I agree totally with that process. The mqtt free should be called after init anyway. I think we are bringing complexity when not needed.

Reply:
But if `mqtt_init` returns something other than `SUCCESS`, users may not know they need to call `mqtt_free`.  I think it makes sense for `init` to clean everything up if it fails.